### ISSUE TITLE
[Bug]: aws_security_group does not plan to add new ingress rule when it points at newly created aws_security_group

### ISSUE THREAD

**Issue by ipc-zpg** [hashicorp/terraform-provider-aws#29564]
### Terraform Core Version

1.3.9

### AWS Provider Version

4.55.0

### Affected Resource(s)

* aws_security_group

### Expected Behavior

Ingress rule with description ="Allow 1431 In - BUG? - will not show in plans" should show in the plan.

### Actual Behavior

Ingress rule with description ="Allow 1431 In - BUG? - will not show in plans" DOES NOT show in the plan.

### Terraform Configuration Files

```terraform
resource "aws_security_group" "new_sec_group" {
  name        = "new_sec_group"
  description = "This sec group does NOT exist yet"
  vpc_id      = module.vpc.vpc_id

  ingress {
    protocol  = -1
    from_port = 0
    to_port   = 0
    cidr_blocks = ["0.0.0.0/0"]
  }

  egress {
    protocol    = -1
    from_port   = 0
    to_port     = 0
    cidr_blocks = ["0.0.0.0/0"]
  }

}

resource "aws_security_group" "sql_server" {
  name        = "sql_server"
  description = "Restricted ingress to SQL server"
  vpc_id      = module.vpc.vpc_id

  ingress {
    protocol  = -1
    description = ""
    from_port = 0
    to_port   = 0
    security_groups = [
      var.vpn_security_group,
      aws_security_group.backup_server.id
    ]
  }

  ingress {
    protocol  = "TCP"
    from_port = 1433
    to_port   = 1433
    description = ""
    security_groups = [
      aws_security_group.default_windows.id,
      aws_security_group.dmz_ftp_windows.id
    ]
  }

  ingress {
    protocol    = "TCP"
    from_port   = 1432
    to_port     = 1432
    description = "Allow 1432 In - existing SG - will show in plans"
    security_groups = [
      aws_security_group.default_windows.id
    ]
  }

  ingress {
    protocol    = "TCP"
    from_port   = 1431
    to_port     = 1431
    description = "Allow 1431 In - BUG? - will not show in plans"
    security_groups = [
      aws_security_group.new_sec_group.id
    ]
  }

  egress {
    protocol    = -1
    from_port   = 0
    to_port     = 0
    cidr_blocks = ["0.0.0.0/0"]
  }

}
```

### Steps to Reproduce

1. Create aws_security_group.sql_server (will need some motifications to remove other resources not included in example - but they have been left in the report so that they match my plan debug)
2. plan the above, which adds aws_security_group.new_sec_group.  The rule allowing 1431 will be missing from plan output.

### Debug Output

https://gist.github.com/ipc-zpg/ff0db0686b3fa70949a186fa224f35ea

### Would you like to implement a fix?

None

**Comment by ipc-zpg** [comment]
Having refactored the code to use attachment rules, rather than embedded rules, TF does actually throw an error:


```
Error: Invalid for_each argument

  on security_group.tf line 224, in resource "aws_security_group_rule" "sql_server_sql_in":
 224:   for_each = toset([
 225:     aws_security_group.dmz_ftp_windows.id,
 226:     aws_security_group.sql_server.id,
 227:     aws_security_group.legacy_admin.id
 228:   ])
    ├────────────────
    │ aws_security_group.dmz_ftp_windows.id is "..."
    │ aws_security_group.legacy_admin.id is a string, known only after apply
    │ aws_security_group.sql_server.id is "..."

The "for_each" set includes values derived from resource attributes that
cannot be determined until apply, and so Terraform cannot determine the full
set of keys that will identify the instances of this resource.

When working with unknown values in for_each, it's better to use a map value
where the keys are defined statically in your configuration and where only
the values contain apply-time results.

Alternatively, you could use the -target planning option to first apply only
the resources that the for_each value depends on, and then apply a second
time to fully converge.
```

In this above, line 227 is an SG added within the same change.

If I don't do anything 'fancy' like doing it in a loop, and have it as one SGR per source_security_group_id , it's fine.
The above suggestion of doing a targeted apply followed by a second apply is nonsense of course - because it would prevent a ground-up rebuild in the future.

So, in summary:
* embedded rules are bad.
* attached rules with a loop though an un-keyed set are bad (example 1 below)
* attached rules with count[] or statically-keyed sets are good. (example 2 below)


```
 resource "aws_security_group_rule" "sql_server_allow_all" {
   security_group_id = aws_security_group.sql_server.id

   for_each = toset([
     var.vpn_security_group,
     aws_security_group.backup_server.id
   ])
   type                     = "ingress"
   protocol                 = -1
   from_port                = 0
   to_port                  = 0
   source_security_group_id = each.value
 }


 resource "aws_security_group_rule" "sql_server_sql_in" {
   security_group_id = aws_security_group.sql_server.id

   for_each = {
     "dmz_ftp" = aws_security_group.dmz_ftp_windows.id,
     "sql_server" = aws_security_group.sql_server.id,
     "legacy_admin" = aws_security_group.legacy_admin.id
   }
   type                     = "ingress"
   protocol                 = "TCP"
   from_port                = 1433
   to_port                  = 1433
   source_security_group_id = each.value
 }
```

**Comment by justinretzolk** [comment]
Hey @ipc-zpg 👋 Thank you for taking the time to raise this! I took a look over the debug logs you supplied, and it looks like you might be targeting the `aws_security_group.new_sec_group` during the plan. Is that correct?

```shell
2023-02-22T08:31:04.652Z [INFO]  CLI args: []string{"terraform", "plan", "-target", "aws_security_group.new_sec_group", "-target", "aws_security_group.sql_server"}
```

If that's correct, I'd not expect to see any changes to `aws_security_group.sql_server`, given the targeting.

**Comment by justinretzolk** [comment]
Since we haven't heard back, I'm going to close this issue. If you're still having trouble, please feel free to open a new issue, referencing this one for context as needed.
